### ISSUE TITLE
Option to pass in memory_mask

### ISSUE THREAD

**Issue by tumbleintoyourheart** [lucidrains/x-transformers#60]
Hi Phil,

Can we have the option to pass in `memory_mask`, just like in [Official Pytorch Transformers](https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html)?

Thanks.

**Comment by lucidrains** [comment]
@tumbleintoyourheart oh hey, yea I have that as well, but it is named "context_mask" instead

**Comment by tumbleintoyourheart** [comment]
@lucidrains Yeah I'm aware of `context_mask`, but I though it is equivalent to `memory_key_padding_mask`, which has shape of (batch_size x context seq_len), in the purpose of masking the padded elements in `context`. Instead I'm talking about `memory_mask`, shape of `query seq_len x context seq_len`, which is broadcasted by `batch_size` and applied to the attention matrix.

What I want to achieve is the same effect of the `Attention masks` in [Figure 4. Architecture of language model (BCN) - paper ABINet (CVPR 2021, Text recognition domain)](https://arxiv.org/pdf/2103.06495.pdf). [The authors' implementation](https://github.com/FangShancheng/ABINet/blob/main/modules/transformer.py) uses Official Pytorch code with some modifications.

Big fan of your works.

**Comment by lucidrains** [comment]
<img width="323" alt="Screen Shot 2021-09-03 at 9 01 26 AM" src="https://user-images.githubusercontent.com/108653/132034626-780b4f6b-1eb2-4d0b-84c3-50f965a83c1f.png">

oh yes, you are correct, so you'd like to mask out the diagonal?

**Comment by tumbleintoyourheart** [comment]
Yeah exactly. Can we achieve that with the current implementation of `mask` and `context_mask`? If not, I think it would be very nice to add the feature.

**Comment by lucidrains** [comment]
@tumbleintoyourheart yup, i can add this, i'll probably name it `attn_mask` - in the presence of memory, i'll just left pad with `True`

**Comment by lucidrains** [comment]
ok, i've deployed it in 0.17.10, let me know if that works for you!

```python
import torch
from x_transformers import XTransformer

model = XTransformer(
    dim = 512,
    enc_num_tokens = 256,
    enc_depth = 1,
    enc_heads = 8,
    enc_max_seq_len = 1024,
    dec_num_tokens = 256,
    dec_depth = 1,
    dec_heads = 8,
    dec_max_seq_len = 1024
)

src = torch.randint(0, 256, (1, 1024))
src_mask = torch.ones_like(src).bool()
tgt = torch.randint(0, 256, (1, 1024))
tgt_mask = torch.ones_like(tgt).bool()

attn_mask = ~torch.eye(1024).bool()

loss = model(src, tgt, src_mask = src_mask, tgt_mask = tgt_mask, src_attn_mask = attn_mask) # (1, 1024, 512)
loss.backward()```

**Comment by tumbleintoyourheart** [comment]
I did some quick tests and it gave instant better results with proper masking thanks to the new feature.

Thanks for your quick responses, it's very nice of you to even write down some sample codes for me. Really appreciated.
